### ISSUE TITLE
Tests fail with llvm 16

### ISSUE THREAD

**Issue by eli-schwartz** [mesonbuild/meson#11642]
See for example https://github.com/mesonbuild/meson/pull/11580#issuecomment-1482108696

```
[UNEXSKIP]   frameworks: 15 llvm    (method=cmake link-static=False)
Reason: Test asked to be skipped (llvm not found.), but was not expected to for MESON_CI_JOBNAME 'macos-clang-unity'
[UNEXSKIP]   frameworks: 15 llvm    (method=cmake link-static=True)
Reason: Test asked to be skipped (llvm not found.), but was not expected to for MESON_CI_JOBNAME 'macos-clang-unity'
  [ERROR]    frameworks: 15 llvm    (method=combination link-static=False)
During: configure
Reason: Generating the build system failed.
  [ERROR]    frameworks: 15 llvm    (method=combination link-static=True)
During: configure
```

Extended logs start here: https://github.com/mesonbuild/meson/actions/runs/4505588849/jobs/7932032771?pr=11580#step:7:954

**Comment by lazka** [comment]
Some fixes here: https://github.com/mesonbuild/meson/pull/11639

**Comment by lazka** [comment]
Reverting https://github.com/llvm/llvm-project/commit/f1d3bc49634cfefe35de3aa631cef9f808b12896 makes things work again for me

**Comment by carlocab** [comment]
Upstream issue: https://github.com/llvm/llvm-project/issues/61436

Fix at https://github.com/llvm/llvm-project/issues/61436#issuecomment-1470517894

I don't really grok CMake, but I wonder if it'll just work to set
```cmake
cmake_policy(SET CMP0057 NEW)
```
in the calling `CMakeLists.txt`.

**Comment by lazka** [comment]
I tired this patch, but it doesn't help here for some reason. Maybe I missed something.

**Comment by carlocab** [comment]
Yea, same: https://github.com/carlocab/meson/actions/runs/4655575421/jobs/8238342973#step:8:1451

Doesn't work for me either. Might be worth mentioning at llvm/llvm-project#61436, but they'll need a minimal reproducer.

**Comment by nirbheek** [comment]
I just tried it — no difference.

```diff
--- a/mesonbuild/dependencies/data/CMakeListsLLVM.txt
+++ b/mesonbuild/dependencies/data/CMakeListsLLVM.txt
@@ -2,7 +2,7 @@
 # cmake_minimum_required(VERSION ${CMAKE_VERSION})
 # project(... LANGUAGES ...)
 
-cmake_policy(SET CMP0000 NEW)
+cmake_policy(SET CMP0057 NEW)
 
 set(PACKAGE_FOUND FALSE)
 
```

**Comment by nirbheek** [comment]
I tried that too, same result

**Comment by rilian-la-te** [comment]
@nirbheek maybe will look in weekend, if ArchLinux have llvm16 in repos.

**Comment by nirbheek** [comment]
Thanks! Arch is still on 15.x, apparently: https://archlinux.org/packages/?sort=&q=llvm&maintainer=&flagged=

You will probably have to try inside a Fedora 38 docker image.

**Comment by rilian-la-te** [comment]
@nirbheek I do not have Docker at home(

**Comment by carlocab** [comment]
You can also do
```
docker run -it homebrew/ubuntu22.04:master /bin/bash
brew install llvm
```
if the Fedora image doesn't work out for some reason. (Warning: Homebrew on Linux is very slow. Sorry.)

**Comment by rilian-la-te** [comment]
@carlocab I do not like the idea of configuring Docker. Maybe I can use CI? Is CI functional on feature branches on forks?

**Comment by carlocab** [comment]
You can use GitHub Actions. Homebrew is installed in the default images. You'll need to add
```
/home/linuxbrew/.linuxbrew/bin
```
to `PATH`, though.

CI should work on forks, but you might need to enable it first in the Actions tab on your fork.

Make sure to use `ubuntu-latest`, `brew` isn't great on older images.

**Comment by thesamesam** [comment]
I can reproduce the issue with this, at least:
```shell
/usr/bin/cmake -DNAME=LLVM -DARCHS= -DVERSION= -DCOMPS= -DSTATIC=True --trace-expand --trace-format=json-v1 --no-warn-unused-cli --trace-redirect=cmake_trace.txt -DLLVM_MESON_REQUIRED_MODULES="bitwriter;asmprinter;executionengine;mcjit;nativecodegen;target;amdgpu;engine" -DLLVM_MESON_OPTIONAL_MODULES= -DLLVM_MESON_PACKAGE_NAMES="LLVM;LLVM-16;LLVM16;LLVM-15;LLVM15;LLVM-14;LLVM14;LLVM-13;LLVM13;LLVM-12;LLVM12;LLVM-11;LLVM11;LLVM-10;LLVM10;LLVM-9;LLVM90;LLVM-8;LLVM80;LLVM-7;LLVM70;LLVM-6.0;LLVM60;LLVM-5.0;LLVM50;LLVM-4.0;LLVM40;LLVM-3.9;LLVM39;LLVM-3.8;LLVM38;LLVM-3.7;LLVM37;LLVM-3.6;LLVM36;LLVM-3.5;LLVM35;LLVM-15;LLVM-devel" -DLLVM_MESON_VERSIONS="90.0;80.0;70.0;60.0;50.0;40.0;39.0;38.0;37.0;36.0;35.0;16.0;16.0;15.0;15.0;15.0;14.0;14.0;13.0;13.0;12.0;12.0;11.0;11.0;10.0;10.0;9.0;8.0;7.0;6.0.0;5.0.0;4.0.0;3.9.0;3.8.0;3.7.0;3.6.0;3.5.0" -DLLVM_MESON_DYLIB=OFF .
```

CMakeLists.txt:
```cmake
cmake_minimum_required(VERSION ${CMAKE_VERSION})
project(MesonTemp LANGUAGES C CXX)
# fail noisily if attempt to use this file without setting:
# cmake_minimum_required(VERSION ${CMAKE_VERSION})
# project(... LANGUAGES ...)

cmake_policy(SET CMP0000 NEW)

set(PACKAGE_FOUND FALSE)

list(REMOVE_DUPLICATES LLVM_MESON_VERSIONS)

while(TRUE)
  #Activate CMake version selection
  foreach(i IN LISTS LLVM_MESON_VERSIONS)
    find_package(LLVM ${i}
      CONFIG
      NAMES ${LLVM_MESON_PACKAGE_NAMES}
      QUIET)
    if(LLVM_FOUND)
      break()
    endif()
  endforeach()

  # ARCHS has to be set via the CMD interface
  if(LLVM_FOUND OR "${ARCHS}" STREQUAL "")
    break()
  endif()

  list(GET       ARCHS 0 CMAKE_LIBRARY_ARCHITECTURE)
  list(REMOVE_AT ARCHS 0)
endwhile()

function(meson_llvm_cmake_dynamic_available mod out)
  # Check if we can only compare LLVM_DYLIB_COMPONENTS, because
  # we do not need complex component translation logic, if all
  # is covered by one variable
  if(mod IN_LIST LLVM_DYLIB_COMPONENTS)
    set(${out} TRUE PARENT_SCOPE)
    return()
  elseif((NOT (mod IN_LIST LLVM_DYLIB_COMPONENTS))
      AND (NOT("${LLVM_DYLIB_COMPONENTS}" STREQUAL "all")))
    set(${out} FALSE PARENT_SCOPE)
    return()
  endif()

  # Complex heuristic to filter all pseudo-components and skip invalid names
  # LLVM_DYLIB_COMPONENTS will be 'all', because in other case we returned
  # in previous check. 'all' is also handled there.
  set(llvm_pseudo_components "native" "backend" "engine" "all-targets")
  is_llvm_target_specifier(${mod} mod_spec INCLUDED_TARGETS)
  string(TOUPPER "${LLVM_AVAILABLE_LIBS}" capitalized_libs)
  string(TOUPPER "${LLVM_TARGETS_TO_BUILD}" capitalized_tgts)
  if(mod_spec)
    set(${out} TRUE PARENT_SCOPE)
  elseif(mod IN_LIST capitalized_tgts)
    set(${out} TRUE PARENT_SCOPE)
  elseif(mod IN_LIST llvm_pseudo_components)
    set(${out} TRUE PARENT_SCOPE)
  elseif(LLVM${mod} IN_LIST capitalized_libs)
    set(${out} TRUE PARENT_SCOPE)
  else()
    set(${out} FALSE PARENT_SCOPE)
  endif()
endfunction()

function(is_static target ret)
  if(TARGET ${target})
    get_target_property(target_type ${target} TYPE)
    if(target_type STREQUAL "STATIC_LIBRARY")
      set(${ret} TRUE PARENT_SCOPE)
      return()
    endif()
  endif()
  set(${ret} FALSE PARENT_SCOPE)
endfunction()

# Concatenate LLVM_MESON_REQUIRED_MODULES and LLVM_MESON_OPTIONAL_MODULES
set(LLVM_MESON_MODULES ${LLVM_MESON_REQUIRED_MODULES} ${LLVM_MESON_OPTIONAL_MODULES})


# Check if LLVM exists in dynamic world
# Initialization before modules checking
if(LLVM_FOUND)
  if(LLVM_MESON_DYLIB AND TARGET LLVM)
    set(PACKAGE_FOUND TRUE)
  elseif(NOT LLVM_MESON_DYLIB)
    # Use LLVMSupport to check if static targets exist
    set(static_tg FALSE)
    is_static(LLVMSupport static_tg)
    if(static_tg)
      set(PACKAGE_FOUND TRUE)
    endif(static_tg)
  endif()
endif()

if(PACKAGE_FOUND)
  foreach(mod IN LISTS LLVM_MESON_MODULES)
    # Reset variables
    set(out_mods)
    set(real_mods)

    # Generate a lower and upper case version
    string(TOLOWER "${mod}" mod_L)
    string(TOUPPER "${mod}" mod_U)

    # Special case - "all-targets" pseudo target
    # Just append all targets, if pseudo-target exists
    if("${mod}" STREQUAL "all-targets")
      set(mod_L  ${LLVM_TARGETS_TO_BUILD})
      string(TOUPPER "${LLVM_TARGETS_TO_BUILD}" mod_U)
    endif()

    # Check if required module is linked is inside libLLVM.so.
    # If not, skip this module
    if(LLVM_MESON_DYLIB
       AND DEFINED LLVM_DYLIB_COMPONENTS)
        meson_llvm_cmake_dynamic_available(${mod} MOD_F)
        meson_llvm_cmake_dynamic_available(${mod_L} MOD_L_F)
        meson_llvm_cmake_dynamic_available(${mod_U} MOD_U_F)
        if(MOD_F OR MOD_L_F OR MOD_U_F)
          set(MESON_LLVM_TARGETS_${mod} LLVM)
        endif()
    elseif(LLVM_MESON_DYLIB AND (mod IN_LIST LLVM_MESON_REQUIRED_MODULES))
      # Dynamic was requested, but no required variables set, we cannot continue
      set(PACKAGE_FOUND FALSE)
      break()
    elseif(LLVM_MESON_DYLIB)
      # Dynamic was requested, and we request optional modules only. Continue
      continue()
    else()
      # CMake only do this for static components, and we
      # replicate its behaviour
      # Get the mapped components
      llvm_map_components_to_libnames(out_mods ${mod} ${mod_L} ${mod_U})
      list(SORT              out_mods)
      list(REMOVE_DUPLICATES out_mods)

      # Make sure that the modules exist
      foreach(i IN LISTS out_mods)
        set(static_tg FALSE)
        is_static(${i} static_tg)
        if(static_tg)
          list(APPEND real_mods ${i})
        endif()
      endforeach()

      # Set the output variables
      set(MESON_LLVM_TARGETS_${mod} ${real_mods})
      foreach(i IN LISTS real_mods)
        set(MESON_TARGET_TO_LLVM_${i} ${mod})
      endforeach()
    endif()
  endforeach()

  # Check the following variables:
  # LLVM_PACKAGE_VERSION
  # LLVM_VERSION
  # LLVM_VERSION_STRING
  if(NOT DEFINED PACKAGE_VERSION)
    if(DEFINED LLVM_PACKAGE_VERSION)
      set(PACKAGE_VERSION "${LLVM_PACKAGE_VERSION}")
    elseif(DEFINED LLVM_VERSION)
      set(PACKAGE_VERSION "${LLVM_VERSION}")
    elseif(DEFINED LLVM_VERSION_STRING)
      set(PACKAGE_VERSION "${LLVM_VERSION_STRING}")
    endif()
  endif()

  # Check the following variables:
  # LLVM_LIBRARIES
  # LLVM_LIBS
  set(libs)
  #Hardcode LLVM, because we links with libLLVM.so when dynamic
  if(LLVM_MESON_DYLIB)
    get_target_property(libs LLVM IMPORTED_LOCATION)
  elseif(DEFINED LLVM_LIBRARIES)
    set(libs LLVM_LIBRARIES)
  elseif(DEFINED LLVM_LIBS)
    set(libs LLVM_LIBS)
  endif()

  # Check the following variables:
  # LLVM_INCLUDE_DIRS
  # LLVM_INCLUDES
  # LLVM_INCLUDE_DIR
  set(includes)
  if(DEFINED LLVM_INCLUDE_DIRS)
    set(includes LLVM_INCLUDE_DIRS)
  elseif(DEFINED LLVM_INCLUDES)
    set(includes LLVM_INCLUDES)
  elseif(DEFINED LLVM_INCLUDE_DIR)
    set(includes LLVM_INCLUDE_DIR)
  endif()

  # Check the following variables:
  # LLVM_DEFINITIONS
  set(definitions)
  if(DEFINED LLVM_DEFINITIONS)
    set(definitions LLVM_DEFINITIONS)
  endif()

  set(PACKAGE_INCLUDE_DIRS "${${includes}}")
  set(PACKAGE_DEFINITIONS  "${${definitions}}")
  set(PACKAGE_LIBRARIES    "${${libs}}")
endif()
```

**Comment by thesamesam** [comment]
Patch at https://reviews.llvm.org/D152121.

**Comment by nirbheek** [comment]
Amazing work, thanks for looking into this!

**Comment by eli-schwartz** [comment]
Thanks for helping me debug this! And the patch is merged.

**Comment by carlocab** [comment]
If there won't be an LLVM 16.0.6 soonish with this fix, I can backport the fix for Homebrew LLVM to help fix your CI here. Otherwise, probably easier to wait.

**Comment by eli-schwartz** [comment]
@thesamesam said that based on the expected schedule the next release would be in 10 days.

**Comment by eli-schwartz** [comment]
@carlocab

According to @thesamesam it appears that there will not end up being a .6 point release.

**Comment by carlocab** [comment]
Ok, that's fine. Will backport. Feel free to ping me every few days until it's done. (But note that I'm quite bad at GitHub notifications.)

**Comment by carlocab** [comment]
Backported in Homebrew/homebrew-core#133601.

**Comment by lazka** [comment]
also backported in MSYS2 now

**Comment by carlocab** [comment]
Merged now along with LLVM 16.0.6.
